cocos/animation/core/transform.ts

```diff
@@ -0,0 +1,228 @@
+import { Vec3 } from '../../core/math/vec3';
+import { Quat } from '../../core/math/quat';
+import { EPSILON, Mat4 } from '../../core';
+
+const CACHE_VECTOR_A = new Vec3();
+const CACHE_VECTOR_B = new Vec3();
+const CACHE_QUAT_A = new Quat();
+const CACHE_QUAT_B = new Quat();
+
+// Can not use `Readonly<Transform>`.
+// See: https://github.com/microsoft/TypeScript/issues/50668
+type ReadonlyTransform = Transform;
+
+export class Transform {
+    public static IDENTITY = Object.freeze(new Transform());
+
+    public static ZERO = Object.freeze((() => {
+        const transform = new Transform();
+        Vec3.copy(transform._position, Vec3.ZERO);
+        Quat.set(transform._rotation, 0.0, 0.0, 0.0, 0.0);
+        Vec3.copy(transform._scale, Vec3.ZERO);
+        return transform;
+    })());
+
+    get position (): Readonly<Vec3> {
+        return this._position;
+    }
+
+    set position (value) {
+        Vec3.copy(this._position, value);
+    }
+
+    get rotation (): Readonly<Quat> {
+        return this._rotation;
+    }
+
+    set rotation (value) {
+        Quat.copy(this._rotation, value);
+    }
+
+    get scale (): Readonly<Vec3> {
+        return this._scale;
+    }
+
+    set scale (value) {
+        Vec3.copy(this._scale, value);
+    }
+
+    public static clone (src: ReadonlyTransform) {
+        const transform = new Transform();
+        Transform.copy(transform, src);
+        return transform;
+    }
+
+    public static setIdentity (out: Transform) {
+        Vec3.copy(out._position, Vec3.ZERO);
+        Quat.copy(out._rotation, Quat.IDENTITY);
+        Vec3.copy(out._scale, Vec3.ONE);
+        return out;
+    }
+
+    public static copy (out: Transform, src: ReadonlyTransform) {
```

Comment:
Why these function use `static` method? I think it looks a bit wired

Reply:
Assignment and copy operations are usually implemented with member methods

Reply:
Other value types in engine are implemented in this way.